diff --git a/src/wolfio.c b/src/wolfio.c
index 0eb5833e3..bd9f6f9f0 100644
--- a/src/wolfio.c
+++ b/src/wolfio.c
@@ -494,13 +494,13 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
     }
     else if (dtlsCtx->userSet) {
         /* Truncate peer size */
-        if (peerSz > sizeof(lclPeer))
-            peerSz = sizeof(lclPeer);
+        if (peerSz > (XSOCKLENT)sizeof(lclPeer))
+            peerSz = (XSOCKLENT)sizeof(lclPeer);
     }
     else {
         /* Truncate peer size */
-        if (peerSz > dtlsCtx->peer.bufSz)
-            peerSz = dtlsCtx->peer.bufSz;
+        if (peerSz > (XSOCKLENT)dtlsCtx->peer.bufSz)
+            peerSz = (XSOCKLENT)dtlsCtx->peer.bufSz;
     }
 
     recvd = TranslateReturnCode(recvd, sd);
