From 9b8dd0de2823cdb023c0c14743024c244d7da813 Mon Sep 17 00:00:00 2001
From: Raihaan Shouhell <raihaan.shouhell@kape.com>
Date: Thu, 20 Feb 2025 13:12:17 +0800
Subject: [PATCH] use proper warnings on MSVC

---
 wolfssl/wolfcrypt/settings.h | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/wolfssl/wolfcrypt/settings.h b/wolfssl/wolfcrypt/settings.h
index e7452de4c..bd49df456 100644
--- a/wolfssl/wolfcrypt/settings.h
+++ b/wolfssl/wolfcrypt/settings.h
@@ -346,7 +346,11 @@
      * an application build -- then your application can avoid this warning by
      * defining WOLFSSL_NO_OPTIONS_H or WOLFSSL_CUSTOM_CONFIG as appropriate.
      */
-    #warning "No configuration for wolfSSL detected, check header order"
+    #if defined(_MSC_VER) /* Microsoft Visual C++ compiler */
+        #pragma message("No configuration for wolfSSL detected, check header order")
+    #else
+        #warning "No configuration for wolfSSL detected, check header order"
+    #endif
 #endif
 
 #include <wolfssl/wolfcrypt/visibility.h>
-- 
2.48.1